### ISSUE TITLE
Terrains UI issues

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3015]
This is several issues but they're all fairly minor and all related to working with Terrains in 1.5. If you'd prefer that I split them up into several issues, I can.

1. When editing a text/number property in the Tileset Editor while in Terrain edit mode, if the cursor leaves the properties panel and moves over a tile (triggering the terrain placement preview), the text/number field loses edit focus. This is inconsistent with the usual behaviour where merely moving the mouse doesn't change focus.

2. In the Map Editor, when switching from the Terrain Brush to another tool and then going back to terrains, it's intuitive to click the terrain you want and then paint, and this is the current behaviour when you click a new terrain. However, if that terrain was already selected, Tiled does not go back to the Terrain Brush and keeps the previous tool selected. I think clicking on a terrain should activate the Terrain Brush, even if the terrain was already selected.

3. When selecting a different terrain set but not selecting a terrain from it, Tiled shows a non-functional preview of the last previewed terrain blob on the side of the editor (the location is probably just due to that being where the preview was last generated). This preview doesn't follow the cursor (probably because there's no terrain to update it with), and clicking to "place" it does nothing. Seems like in this situation, the cursor being in the editor causes the preview to be shown, but not to be updated. Since the Erase Terrain button is highlighted after switching terrains, it looks like that's the intended behaviour? But the erase terrain mode is not actually activated.
![tiledterrainglitch](https://user-images.githubusercontent.com/70545646/112768784-b6175100-8feb-11eb-8cd3-b35f0895baf1.gif)

4. It's possible for the Terrains list in the Terrain Sets panel to be collapsed. Does it ever make sense for the terrains list to be collapsed while the Terrain Sets panel is in use? A collapsed terrains list means users can't even use the Terrain Brush until they expand it and select a terrain, since the Erase Terrain mode is selected by default when they select a Terrain Set. When it is collapsed, it's also hard to see that it's even there. I think the terrains list should be resizable but not collapsible. This goes for both the Map Editor and the Tileset Editor. The terrains list is just too important for use of the feature, and causes a lot of confusion when it's unintentionally collapsed and the user isn't already aware that it *should* be there.

5. On a related note to (4), it's entirely possible for the user to have the entire Terrain Sets panel hidden (whether by default or by accident) while trying to use the Terrain Brush, causing the tool to do nothing. Users not already familiar with how the tool works may not know that they need to switch to or enable this panel. I don't have a good solution for this one. Opening the panel for them or raising a Warning punishes quick misclicks with possibly unwanted clutter and warnings. A modal warnings with the option to open the panel seems like overkill. Perhaps a solution could be to open the panel, *but* to also disable the tool when none of the map's tilesets have terrains, to prevent misclicks? People who do have terrains are likely to have the panel enabled already anyway, or to want it enabled.

**Comment by bjorn** [comment]
> 1. When editing a text/number property in the Tileset Editor while in Terrain edit mode, if the cursor leaves the properties panel and moves over a tile (triggering the terrain placement preview), the text/number field loses edit focus. This is inconsistent with the usual behaviour where merely moving the mouse doesn't change focus.

Hmm, this was introduced in #1638, probably to make the keyboard shortcuts work. For now I've just removed the focus change, but it would be nice to still have those shortcuts work even if focus isn't in the tileset view. The shortcuts also turned out to be non-functional though, but I fixed that in 23da9131e3ed7cf6df159528f027ecf95ad91424.

> 2. ... However, if that terrain was already selected, Tiled does not go back to the Terrain Brush and keeps the previous tool selected. I think clicking on a terrain should activate the Terrain Brush, even if the terrain was already selected.

Fixed in 68bd2a2cea9352551d41a03aa377703336062e53.

> 4. It's possible for the Terrains list in the Terrain Sets panel to be collapsed. Does it ever make sense for the terrains list to be collapsed while the Terrain Sets panel is in use?

I don't think so, made not collapsible in ab6541c72ff89c0a948bd25fc0856ba9aad94056.

Thanks for the feedback! I'll have a look at 3 and 5 tomorrow.

**Comment by bjorn** [comment]
I've improved upon 3 by automatically selecting the first terrain (07689a581b7d61657a55dfecaab92d91657a890e).

Regarding 5, as discussed on Discord I didn't manage to find a nicer solution to automatically hiding/showing this view for now. In 07a38d77baa72d6aa4a14fd8715b123a6b4fddec I did make the Terrain Sets view a little more helpful in case there isn't any tileset with a terrain set:

![image](https://user-images.githubusercontent.com/531764/114887705-fad32280-9e08-11eb-9646-39c8cd98acf2.png)

Thanks again for your feedback! I think we can close this issue now. :-)
